docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,205 @@
+
+### Model Prompts
+
+Based on whether model_family is mixtral or merlinite, a different prompt is used with the teacher model
+
+```python
+_MODEL_PROMPT_MIXTRAL = "<s> [INST] {prompt} [/INST]"
+_MODEL_PROMPT_MERLINITE = "'<|system|>\nYou are an AI language model developed by IBM Research. You are a cautious assistant. You carefully follow instructions. You are helpful and harmless and you follow ethical guidelines and promote positive behavior.\n<|user|>\n{prompt}\n<|assistant|>\n'"
+```
+
+For now, we assume that the `LLMBlock` class will choose the appropriate model prompt based on the family and that there is no need to specify a custom prompt.
+
+### Prompt Config Files
+
+Every LLMBlock references a separate prompt config file, and presumably a custom pipeline will provide custom prompt configs too.
+
+These prompt config files are quite simple YAML files - they contain a single object with system, introduction, principles, examples, and generation keys. See e.g. src/instructlab/sdg/configs/skills/freeform_questions.yaml
+
+We will continue to use these config files unchanged, and custom files can be specified with an absolute path. Relative paths are assumed to be relative to the Python package e.g. `configs/skills/...`.
+
+### Model Serving
+
+Custom pipelines may have more unique model serving requirements. Instead of serving just one model, we may need to launch the model server with a model and an additional model with adapter. vLLM, for example, can host both a model and a model+adapter under two different model IDs.
+
+The pipeline author needs some way of disambiguating between these multiple models - i.e. the definition of each `LLMBlock` needs to specify a particular model.
+
+Right now the `Pipeline` constructor takes two relevant parameters - the OpenAI client instance, and the model ID for the default model. It's important to note that this model ID is defined by the user at runtime, and it may not match the model IDs that the pipeline author used.
+
+The use cases will be:
+
+1. Most LLMBlock definitions will use the default teacher model - and we can make the semantic that if the pipeline author doesn't specify a model in an `LLMBlock`, the default in `PipelineContext.model_id` is used.
+2. In cases where a model+adapter is to be served, the pipeline author should choose a descriptive model ID using `block.gen_kwargs.model_id` and the user should ensure that this is the model ID that is served.
+
+For example, a pipeline author might define:
+
+```yaml
+version: "1.0"
+blocks:
+  - name: gen_questions
+    type: LLMBlock
+    config:
+      config_path: configs/skills/freeform_questions.yaml
+      add_num_samples: True
+      gen_kwargs:
+        model_id: mycustomadapter
+      output_cols:
+        - question
+    drop_duplicates:
+      - question
+```
+
+and the user will be required to define a serving configuration like:
+
+```bash
+--lora-modules=mycustomadapter=path/to/my_custom_adapter
+```
+
```

Comment:
```suggestion

### Re-use of Built-in Pipelines

A custom pipeline may want to add to an existing built-in pipeline.  In that case, in the list of blocks, one may specify a special entry that indicates what to inherit from.

```yaml
version: "1.0"
blocks:
  - inherit_from: configs/full/freeform_skills.yaml
  - name: gen_questions
    type: LLMBlock
    ...
```

Note that an `inherit_from` entry can appear in any location in the `blocks` list (beginning, somewhere in the middle, end).

Reply:
a more "pure" approach here could be a new block type, like "CallBlock" that acts like a function call of sorts, calling out to a separate pipeline definition and then returning back here .

... but that feels like overkill right now maybe?

Reply:
ChatGPT [suggested](https://chatgpt.com/share/7a6f323e-d190-4a19-b42c-c56d8403b93d) YAML Anchors, Aliases and Merge Keys ... but I don't think it's a great fit

How about "import" instead of "inherit"?

And a future expansion might look like:

```yaml
version: "1.0"
blocks:
  - import:
      - path: configs/full/freeform_skills.yaml
        blocks:
         - gen_questions
         - eval_questions
      - path: ...
  - name: gen_questions
    type: LLMBlock
    ...
`` `

(we would add the "import blocks" bit now, that would be a future enhancement potentially)

Reply:
ok, so I can capture this as stages.  A shorter term ` - import_from: path/to/yaml` and then a longer term nicer expansion that's more expressive like you laid out the example.

I'll capture it like that in a doc update

Reply:
ah ha!  thank you.

Reply:
Bleh, no - I think you were right - an import should be another type of block

```
version: 1.0
blocks:
  - name: import_1
     type: ImportBlock
     path: 1.yaml
```

plenty of room for expansion while still respecting that the blocks clause is a list of block objects

Reply:
OK - yeah ... everything else is a hack and a special case, where this seems "right"

it's probably not any harder to implement ... i hope ...

Reply:
Looks good to me. The `ImportBlock` is a neat idea!

Reply:
I pushed an update to reflect the use of an `ImportBlock` to reference another pipeline configuration file.